Adds support for a custom SymbolProvider in NativeLibrary & Library

soywiz opened the pull request:
Retake on https://github.com/java-native-access/jna/pull/1483 I'm using a different branch, since I used master and that was problematic for syncing the original repo. I have also squashed all the commits into one.

---

This allows to do function hooking/replacement while still using [direct mapping](https://github.com/java-native-access/jna/blob/master/www/DirectMapping.md).

In addition to that, to properly using the OpenGL library on Windows, and access OpenGL >= 2.0, we need to use a custom `getSymbol` function: https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-wglgetprocaddress

In the test I'm switching `cos` and `sin` functions to show it works.

---


Started a thread here: https://groups.google.com/g/jna-users/c/RDt8sHAnTm0

soywiz commented:
@matthiasblaesing can this be considered? is there something else I can do to move this forward?

soywiz commented:
Sure. No problem and no rush. Just wanted to know if I missed something. Thanks for letting me know!

matthiasblaesing reviewed (changes requested):
Looks ok in general to me. What I noticed is:

- This change also affects the `InvocationHandler` codepath, so this also needs to have a test for the non-direct invocation case.
- The naming of the third parameter in `SymbolProvider#getSymbolAddress` makes me wonder if `parent` is really right here. Though I have not really reached a good conclusion
- Please recheck the commit and ensure, that your real name is in the author field (`git log -n 1` is your friend)

For the change in `NativeLibrary` - what do you think about this:

```diff
--- a/src/com/sun/jna/NativeLibrary.java
+++ b/src/com/sun/jna/NativeLibrary.java
@@ -87,17 +87,22 @@
 public class NativeLibrary implements Closeable {
 
     private static final Logger LOG = Logger.getLogger(NativeLibrary.class.getName());
-    private final static Level DEBUG_LOAD_LEVEL = DEBUG_LOAD ? Level.INFO : Level.FINE;
+    private static final Level DEBUG_LOAD_LEVEL = DEBUG_LOAD ? Level.INFO : Level.FINE;
+    private static final SymbolProvider NATIVE_SYMBOL_PROVIDER = new SymbolProvider() {
+        @Override
+        public long getSymbolAddress(long handle, String name, SymbolProvider parent) {
+            return Native.findSymbol(handle, name);
+        }
+    };
 
     private Cleaner.Cleanable cleanable;
     private long handle;
     private final String libraryName;
     private final String libraryPath;
     private final Map<String, Function> functions = new HashMap<String, Function>();
+    private final SymbolProvider symbolProvider;
     final int callFlags;
     private String encoding;
-    private SymbolProvider symbolProvider;
-    private SymbolProvider defaultSymbolProvider;
     final Map<String, ?> options;
 
     private static final Map<String, Reference<NativeLibrary>> libraries = new HashMap<String, Reference<NativeLibrary>>();
@@ -125,15 +130,11 @@
         this.callFlags = callingConvention;
         this.options = options;
         this.encoding = (String)options.get(Library.OPTION_STRING_ENCODING);
-        this.symbolProvider = (SymbolProvider)options.get(Library.OPTION_SYMBOL_PROVIDER);
-        if (this.symbolProvider != null) {
-            //noinspection Convert2Lambda
-            defaultSymbolProvider = new SymbolProvider() {
-                @Override
-                public long getSymbolAddress(long handle, String name, SymbolProvider parent) {
-                    return Native.findSymbol(handle, name);
-                }
-            };
+        SymbolProvider optionSymbolProvider = (SymbolProvider)options.get(Library.OPTION_SYMBOL_PROVIDER);
+        if (optionSymbolProvider == null) {
+            this.symbolProvider = NATIVE_SYMBOL_PROVIDER;
+        } else {
+            this.symbolProvider = optionSymbolProvider;
         }
 
         if (this.encoding == null) {
@@ -649,11 +650,7 @@
         if (handle == 0) {
             throw new UnsatisfiedLinkError("Library has been unloaded");
         }
-        if (symbolProvider != null) {
-            return symbolProvider.getSymbolAddress(handle, name, defaultSymbolProvider);
-        } else {
-            return Native.findSymbol(handle, name);
-        }
+        return this.symbolProvider.getSymbolAddress(handle, name, NATIVE_SYMBOL_PROVIDER);
     }
 
     @Override

```

soywiz commented:
@matthiasblaesing applied your patch since it makes sense to me. Also added my real name in the git commit.

Regarding to "`InvocationHandler` codepath" I don't know what that really means. I mean: for me there are two code paths, the default one, and the one that provides a custom `OPTION_SYMBOL_PROVIDER`. The default one was already being tested, and the new one is also tested in the PR. And I can't see any `InvocationHandler` reference in the PR, so I think I need a small clarification on that.

matthiasblaesing commented:
> @matthiasblaesing applied your patch since it makes sense to me. Also added my real name in the git commit.
> 
> Regarding to "`InvocationHandler` codepath" I don't know what that really means. I mean: for me there are two code paths, the default one, and the one that provides a custom `OPTION_SYMBOL_PROVIDER`. The default one was already being tested, and the new one is also tested in the PR. And I can't see any `InvocationHandler` reference in the PR, so I think I need a small clarification on that.

I mean this construct (untested!):

```java
    interface MathInterfaceWithSymbolProvider extends Library {

        public MathInterfaceWithSymbolProvider INSTANCE = Native.load(
                Platform.MATH_LIBRARY_NAME,
                MathInterfaceWithSymbolProvider.class, Collections.singletonMap(
                        Library.OPTION_SYMBOL_PROVIDER,
                        new SymbolProvider() {
                            @Override
                            public long getSymbolAddress(long handle, String name, SymbolProvider parent) {
                                if (name.equals("sin")) {
                                    return parent.getSymbolAddress(handle, "cos", null);
                                } else {
                                    return parent.getSymbolAddress(handle, "sin", null);
                                }
                            }
                        }
                )
        );

        double cos(double x);
        double sin(double x);
    }
```

`Native#load` takes an interface, that defines the mappings to the native functions as abstract methods. For this interface a [java.lang.reflect.Proxy](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/Proxy.html) is created, which delegates invocations to an ([java.lang.reflect.InvocationHandler](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/package-summary.html)).

The implementation can be found here: https://github.com/java-native-access/jna/blob/9e7160e57a394ac650d5151a79d1703903173de8/src/com/sun/jna/Library.java#L117-L268

HTH
Matthias

soywiz commented:
I see. So I guess incidencially this should be usable with the interface variant too. I have updated the title of the PR, updated the README and added a test

matthiasblaesing reviewed (commented):
Final nitpick: Please move the tests together into `NativeCustomSymbolProviderTest`. That way the test is focus on a single feature and not spread over two classes.

soywiz commented:
@matthiasblaesing Done.
This test failed: https://app.travis-ci.com/github/java-native-access/jna/builds/258975707#L1439-L1443
though it looks unrelated, since that tests tries to write to invalid memory to figure out if an exception is thrown.
Is there anything else I can do here?

matthiasblaesing reviewed (approved):
Change looks good to me and Appveyor, Travis and github actions agree. Thank you.
